lib/request.js

```diff
@@ -79,9 +79,14 @@ exports = module.exports = internals.Request = class {
     static generate(server, req, res, options) {
 
         const request = new internals.Request(server, req, res, options);
+        const proto = Object.getPrototypeOf(request);
 
-        // Decorate
+        // Give request access to all decorations without adding them to
+        // the Request prototype.
+        Object.setPrototypeOf(request, server._core._decorations.request);
+        Object.setPrototypeOf(server._core._decorations.request, proto);
```

Comment:
Are there upsides to this versus the approach taken with `requestApply`?

Reply:
You mean looping over the decorations? There shouldn't be. Depending on the number of decorators, this might be faster since you're just adjusting some pointers vs. looping. I mostly went with this approach since it seemed more similar to the existing implementation (in case code was relying on decorations being on the prototype).

Reply:
Ah, that makes a lot of sense.  I'm a little concerned by the performance warning MDN gives about [`setPrototypeOf()`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/setPrototypeOf) (which I can't personally confirm or deny for current-day V8).  But also, setting the prototype of `server._core._decorations.request` to the same prototype on each request seems like it may be avoidable.  Here's one idea for an alternative implementation that might have less question marks re: performance.  https://github.com/hapijs/hapi/compare/master...devinivy:3718-alternative

Reply:
I like that approach better. I'm going to close this.

Reply:
I made a PR taking that approach in #3796—feel free to suggest improvements.  I nabbed the test you had written here—if you'd like a proper "commit" credit for that, I'm happy to take the time.